src/schema/validator.js

```diff
@@ -19,183 +23,391 @@ import {
 } from './formats';
 import schemas from './imported';
 
-const validator = ajv({
-  allErrors: true,
-  errorDataPath: 'property',
-  jsonPointers: true,
-  verbose: true,
-  schemas,
-  schemaId: 'auto',
-});
-
-validator.addMetaSchema(require('ajv/lib/refs/json-schema-draft-06.json'));
-
-ajvMergePatch(validator);
-
-validator.addFormat('versionString', isValidVersionString);
-validator.addFormat('contentSecurityPolicy', () => true);
-validator.addFormat('ignore', () => true);
-validator.addFormat('manifestShortcutKey', manifestShortcutKey);
-
-// URL formats. The format names don't mean what you'd think, see bug 1354342.
-//
-// url -> MUST be absolute URL
-// relativeUrl -> CHOICE of absolute URL or relative URL (including protocol relative)
-// strictRelativeUrl -> MUST be relative, but not protocol relative (path only)
-validator.addFormat('url', isAbsoluteUrl);
-validator.addFormat('relativeUrl', isAnyUrl);
-// homepageUrl is the same as relativeUrl but Firefox will encode | characters.
-validator.addFormat('homepageUrl', isAnyUrl);
-validator.addFormat('strictRelativeUrl', isStrictRelativeUrl);
-validator.addFormat('unresolvedRelativeUrl', isUnresolvedRelativeUrl);
-validator.addFormat('secureUrl', isSecureUrl);
-
-validator.addFormat(
-  'imageDataOrStrictRelativeUrl',
-  imageDataOrStrictRelativeUrl
-);
-
-validator.addKeyword('deprecated', {
-  validate: function validateDeprecated(message, propValue, schema, dataPath) {
-    if (
-      !Object.prototype.hasOwnProperty.call(
-        DEPRECATED_MANIFEST_PROPERTIES,
-        dataPath
-      )
-    ) {
-      // Do not emit errors for every deprecated property, as it may introduce
-      // regressions due to unexpected new deprecation messages raised as errors,
-      // better to deal with it separately.
-      return true;
-    }
-
-    validateDeprecated.errors = [
-      {
-        keyword: 'deprecated',
-        message,
-      },
-    ];
-
-    return false;
-  },
-  errors: true,
-});
+const jsonSchemaDraft06 = require('ajv/lib/refs/json-schema-draft-06.json');
 
 function filterErrors(errors) {
   if (errors) {
-    return errors.filter((error) => error.keyword !== '$merge');
+    return errors.filter((error) => {
+      return error.keyword !== '$merge';
+    });
   }
   return errors;
 }
 
-const _validateAddon = validator.compile({
-  ...schemaObject,
-  id: 'manifest',
-  $ref: '#/types/WebExtensionManifest',
-});
+export class SchemaValidator {
+  // TODO: add jsdoc with the expected validator options.
+  constructor(validatorOptions) {
+    this._options = validatorOptions;
```

Comment:
nit: do we need underscore prefixes?

Reply:
I can, it is more of a warning for us to not access or change that from outside the class (me looking forward to start using the standardized private class fields instead of this kind of "note to self" conventions :-)).

we have a bunch of similar cases like this in this repo (I didn't look to closely, I have just grepped a bit using `egrep "\._" src -r`), if you feel strongly about it I'm definitely ok on removing the _ from this class, but it may as well be reasonable to keep following the same convention we already used elsewhere in this project and then change it consistently across the entire code base when we decided how we like to handle this in the long run (e.g. from a quick look it seems that nodejs 12 may have introduced support for the private fields, if that is enabled by default without any cli flag we may look into migrating to that).

But... I did notice a TODO here related to describing what validatorOptions is expected to contain, I'll take care of that in the meantime.

Reply:
I don't really care. I think I tried to ban that convention over time but not in all repos yet obviously :p If this convention exists in this repo, and you want to keep it here, it's fine.